### ISSUE TITLE
new buffer library

### ISSUE THREAD

**Issue by samoht** [mirage/mirage#881]
The current situation with buffers is a bit annoying. The various MirageOS (and other) libraries are using, at lease: `strings`, `bytes`, `bigarrays` and `Cstruct.t`.

Because of that, there are often a lot of unnecessary copie, especially between `bytes` and `Cstruct.t`. Four libraries are not enough,  so it is probably time to develop a new standard :-)

More seriously, I think we came to the conclusion that we need something between `bytes` (lightweight buffer, not much memory pressure) and `Cstruct.t` (unmovable memory segment, cheap substrings): so I propose we start developing something:
- compatible with `bytes` but without managed by the GC (so not moved around), useful for security and C bindings
- keep the same API as `Cstruct` (even if it could be rationalised a bit)
- could be using something like ancient, e.g. allocate a big chunk of memory and manage its memory itself, or something more clever.

Any thoughts?

**Comment by hannesm** [comment]
yes!  this is a great idea!  what I'm curious: couldn't we keep the cstruct API, but replace the backend to use bigstring/string/bytes instead of bigarray?  and in case someone is doing it from scratch, it would _really_ be useful to have the readonly / readwrite distinction at the type level (as string vs bytes does atm, as https://github.com/mirage/ocaml-cstruct/issues/14 suggested 5 years ago ;)

I talked back at icfp 2017 with @seliopou about his decisions to remove cstruct from angstrom and faraday -- I think it was "less dependencies" and "more performance".

imho cstruct does a bit too many checks upon access, my use pattern is usually:
`buf` was received somehow, now I parse it.  I ensure _at the beginning_ that the length is sufficient for the header, now retrieving the bits of the header (let's say `16 * uint8`) leads to lots of index checks (`i >= 0 && i <= buf.len`) -- which are superfluous -- it would be great to have this in mind (and provide a nicer interface with fewer checks) when designing such a new library.  also some thought may need to be done on unsigned and signed integer types at the high level (i.e. providing nice ways to interact with ocaml-integers and/or usane -- but keeping the new cstruct mostly dependency-free would be great as well).

I'd claim that "buffer-aligned, non-moving memory" can still be optional, i.e. the io-page library using cstruct.t (and I use cstruct.t for lots of things which does not need to be aligned, I usually don't care about moving).

cryptographic secrets: this is complicated, since you'll need to (taint-)track derived secrets (see https://blogs.akamai.com/2014/04/heartbleed-update-v3.html where they discovered that while p and q were kept secret, p-1 and q-1 were not).  prior work on ocaml-sodium regarding "keeping things secret" at https://github.com/klakplok/ocaml-sodium/commit/6d05c65a4d9512dff98db4a838e28e3a6dfbc841

**Comment by samoht** [comment]
Thanks. Just to clarify why non-moving is useful: it helps with interfacing with C code, as you know that the GC will not move the objects behind your back. I know they are other ways to achieve this, but using bigarrays for this is usually quite convenient.

**Comment by hannesm** [comment]
@samoht hmm, does this mean that if I pass a `bytes` or `string` to a C function, this may eventually break due to GC moving stuff around? (to me, http://caml.inria.fr/pub/docs/manual-ocaml/intfc.html sounds more that once in C land `CAMLparamX` is used, this informs the GC to deal with it).

I understand that using Bigarray is at least for direct (page-aligned) communication with hardware (xen, DMA, PCI, other busses and drivers) required.  But since allocation of a Bigarray is kind of expensive, code like https://github.com/mirleft/ocaml-tls/blob/0.9.0/lib/writer.ml is best to be avoided (lots of `Cstruct.create n` with `0 < n <= 5` -- each assembly of a TLS frame requires tons of Bigarray allocations)!?  -- as alternative, allocate once and pass the buffer and offset around, should be less overhead (as done in https://github.com/roburio/udns/blob/c08ed27e04248b93476258214457d39fcf5a8e5b/src/dns_packet.ml#L928) -- this works fine if there's a notion of maximum packet size (in µDNS I allocate 512 bytes foreach outgoing packet -- this is also not optimal for two reasons: (a) TCP and TLS transport can transport more (b) most DNS frames are much smaller than 512 byte).

**Comment by haesbaert** [comment]
I think the real issue with cstruct is always in the writer front, parsing
from it is usually ok.

I had this issue in awa-ssh where I usually want to append data to a
Cstruct.t, but it's a lot of small appends, and I don't it each one of them
to be a malloc+copy.
So I've built this wrapper around it, where it allocates a big chunk, and
resizes when needed, avoiding the reallocation and copy:

https://github.com/haesbaert/awa-ssh/blob/master/lib/dbuf.ml

It starts with a 2k chunk, and grows another 2k if needed, this way the
user can have a quite readable idiom when assembling a packet:

| Msg_channel_open_failure (recp_channel, reason, desc, lang) ->
put_id MSG_CHANNEL_OPEN_FAILURE buf |>
put_uint32 recp_channel |>
put_uint32 reason |>
put_string desc |>
put_string lang

**Comment by hannesm** [comment]
@haesbaert
>  I think the real issue with cstruct is always in the writer front, parsing from it is usually ok.

I disagree. As mentioned above, lots of unneeded range checks (which are conditionals / branches) in a common use case. IIRC the angstrom measurements, the performance (time) impact was nearly a factor of two between raw bigarray and cstruct.

> So I've built this wrapper around it, where it allocates a big chunk, and resizes when needed, avoiding the reallocation and copy:

That sounds interesting, but bears a DoS issue: when the attacker can send you a single byte, and you allocate 2k, that just sounds wrong (in ssh it may not be relevant since you need an established TCP flow -- but for implementing TCP (or DNS or sth via UDP), that should better be avoided).  I'm not convinced that pre-allocation is a good strategy, but rather believe in [finger trie](https://en.wikipedia.org/wiki/Finger_tree).

**Comment by haesbaert** [comment]
@hannesm

On the DoS. I disagree, you have to know what you’re doing, there is no
escape. Every network stack allocates bigger chunks than necessary for
sending a byte. Normally you would have one chunk per unit of work and
raise if necessary, if you have two, something is wrong. Sorry but this is
just FUD, it’s like me saying the attacker can force you to make many small
allocations and then fragment your memory, well if you allow that, yes he
can.

Your example is the other way around, I was talking about the sending path,
not the receiving, you only use Dbuf for reating data to send, not teceive.

Also, if you rceive 1 byte, you already have a page allocated in the card
buffer, the problem of allocating 2k per byte is only a problem well, if
you can’t control the 2k per byte, in other words, if you can create 2k
buffers without a boundary (without getting rid of the previous 2k).

On the performance of reading side, so if you remove the checks from
cstruct, you get twice the performance ? Or it’s just an assumotion that
the checks are the issue ?

**Comment by seliopou** [comment]
In addition to performance and dependencies, the cstruct write API is indeed also not great. My solution to this specifically in the write side is [faraday](https://github.com/inhabitedtype/faraday), which not only makes writing to a single big string more ergonomic but also manages a queue of bigsgrings to be written for you. That's nice since if your output ultimately has to be sent to a system call, you can use a "vectorized" version of it (i.e., `writev` versus `write`) if you happen to have scheduled multiple buffers yourself, or because faraday had to allocate multiple buffers internally to satisfy your writes (and kept the original buffer around rather than doing a copy).

I've gotten similar feedback about the API by default not allocating for small writes, but eliminating that possibility doesn't seem like the right foundation to build. If you expose a way to query the remaining write space, then you can build an allocation-safe version on top of it.

**Comment by hannesm** [comment]
(a) discussion of `small bigarray allocations` at https://github.com/ocaml/ocaml/pull/1723#issuecomment-381616858
(b) @seliopou now moved the low level bits into https://github.com/inhabitedtype/bigstringaf which may be useful as abstraction below cstruct (or replacing them?)

**Comment by cfcs** [comment]
1) FWIW this is what I use: https://github.com/cfcs/ocaml-cs/blob/master/lib/cs.mli

I have come to like the `R` and `W` modules more and more, and the pretty-printer for `R` has come in handy many times. They keep an internal offset pointer to enable streaming read/write.
Example: https://github.com/cfcs/ocaml-luks/blob/master/lib/luks.ml#L67-L100

2) The performance-oriented people are probably not going to like the use of `rresult` vs exceptions, but I really hate undocumented exceptions. The library grew kind of organically out of things I wanted, so naming and types are a bit inconsistent, but I also found it useful to expose exception-throwing functions.

3) Re: backing: small bigarray allocations are very costly, I think everyone agrees.

4) I think phantom typing would be very useful for something like:
```ocaml
val get_char : (foo: [< `R | `W | `RW] cstruct) -> int64 -> (char, 'error) result

val set_char : (foo: [< `W | `RW ] cstruct) -> int64 -> char -> (char, 'error) result
```

5) I could imagine a similar thing for the backing storage (most functions are probably not going to care where the memory lives). For things you want to give to C functions you could have a
```ocaml
val do_something_in_c : (foo : ([< `W | `RW] * [ `Page_aligned_storage ]) cstruct -> bool

val do_something_in_ocaml : (foo : ([< `W | `RW] * 'whatever) cstruct -> bool
```

6) I have a gut feeling this would be an excellent time to discuss copy-on-write cstructs as well, which for example could be implemented as a backing type, and would allow functions that converted from ```[`R]``` to ```[`RW]```, but skipping the actual `memcpy` if nothing is changed.

7) Another thing I think is worth discussing is that with the current design a `Cstruct.t` can change "underneath" you, meaning you can run into time-of-check-vs-time-of-use bugs like if you have an Lwt-function where you first validate the contents, then yield, then do something with the contents (that were changed after your yield). It might make sense to distinguish between this sort of *read-only, but others can write to it* vs *really immutable*

**Comment by hannesm** [comment]
thanks @cfcs.  I like copy-on-write as well!

- bytes/string vs bigarray: @samoht agreed, interfacing with C is crucial.  I'd be ok with cfcs (5) to retrieve a (potentially) page-aligned buffer out of a cstruct for passing to C
- immutable cstructs: I'd like to have these, e.g. for the receiving side of data from network/block - I don't want anything to be able to mutate
- this seems to be interaction between (a) ownership (if there's no alias handed out (we're the `unique` owner), we can do anything to the cstruct without loosing control), (b) mutability (`read-only` / `read-write` / `write-only`), and (c) preemption / atomic operations:
   - a `unique` cstruct is safe to be modified in any regards
   - a `shared` cstruct which is `read-only` for everybody is referential transparent
   - a `shared` cstruct which is `write-only` for everybody is behaving nicely as well (since nobody can observe the modification)
   - a `shared` cstruct where someone has `read-write` or `write-only` handles is not referential transparent and thus tricky to deal with
   - now if we've a cooperative multitasking world, we can rely on not being yielded before the block.  in case we're running with Lwt_preemptive tasks, we can't.  on MirageOS there's no preemption \o/

sounds very much like some linearity requirements and borrow checker would be nice to have here - I'm curious how much of this can be encoded into the OCaml type system already (maybe some features have to wait for the effect system).

**Comment by hannesm** [comment]
this https://discuss.ocaml.org/t/a-proposal-for-a-resource-management-model-for-ocaml/1680 looks related to my thoughts

**Comment by cfcs** [comment]
Not sure if this is a good place to bring this up, but for the non-ocaml-heap-allocated (bigarray-style) allocations it would also be cool to be able to specify that the buffer should be overwritten before being freed. Being unable to reliably get rid of sensitive data is a problem that comes up again and again when dealing with things like encryption keys, and we're not going to be able to solve it completely, but maybe this is a chance to take a step towards that, and the features from the paper @hannesm linked to might also be able to get us closer. Just a random thought.
